quantumblacklabs/kedro-viz
# [KED-2528] Improve layer solving approach when layers partially defined

## Description

This PR improves the approach for inferring and rendering nodes in pipelines with only partially defined layers.

## Development notes

The algorithm for nearest layer search was changed to strictly follow rank order, avoiding edge cases where layer constraints could not be solved with the previous algorithm.

The strict constraint solver was also changed to allow partial solving whilst throwing a console warning, though this was no longer observed to happen in practice.

## QA notes

Tested using a large pipeline that has only partially defined layers (i.e. not the 'demo').
The 'demo' dataset, which has fully defined layers, was also tested.

Using both cases above, open the console to observe any errors or warnings, then test by enabling layers and toggling various combinations of nodes as well as toggling top level pipelines in various combinations. 

Each time check that the resulting render is in a sensible order, does not appear broken and there are no console errors or warnings. Since the order is inferred using only limited information, it may not always be exactly as desired but in this case the idea is that the user should add more layer information to nodes.

## Checklist


## Legal notice


- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
